From e5ed0a3fb2571d69436bb9bc3bcaad4d60878637 Mon Sep 17 00:00:00 2001
From: Yu-Ming Chang <yumin686@andestech.com>
Date: Wed, 16 Aug 2023 16:16:26 +0800
Subject: [PATCH 040/241] target/riscv: Add CSR support of mvec_cfg, mfiob, and
 ucode. Signed-off-by: Yu-Ming Chang <yumin686@andestech.com>

---
 target/riscv/csr_andes.c | 79 ++++++++++++++++++++++++++++++++++++----
 target/riscv/csr_andes.h |  4 ++
 2 files changed, 75 insertions(+), 8 deletions(-)

diff --git a/target/riscv/csr_andes.c b/target/riscv/csr_andes.c
index 054017e57f..79bdc80458 100644
--- a/target/riscv/csr_andes.c
+++ b/target/riscv/csr_andes.c
@@ -49,6 +49,17 @@ static RISCVException ecd(CPURISCVState *env, int csrno)
     }
 }
 
+static RISCVException edsp(CPURISCVState *env, int csrno)
+{
+    AndesCsr *csr = &env->andes_csr;
+    if (get_field(csr->csrno[CSR_MMSC_CFG], MASK_MMSC_CFG_EDSP) == 0) {
+        return RISCV_EXCP_ILLEGAL_INST;
+    }
+    else {
+        return RISCV_EXCP_NONE;
+    }
+}
+
 static RISCVException pft(CPURISCVState *env, int csrno)
 {
     AndesCsr *csr = &env->andes_csr;
@@ -88,11 +99,38 @@ static RISCVException ppma(CPURISCVState *env, int csrno)
     if (get_field(csr->csrno[CSR_MMSC_CFG], MASK_MMSC_CFG_PPMA) == 0) {
         return RISCV_EXCP_ILLEGAL_INST;
     }
+    else {
+        if (csrno == CSR_PMACFG1 || csrno == CSR_PMACFG3) {
+            if (riscv_cpu_mxl(env) != MXL_RV32) {
+                return RISCV_EXCP_ILLEGAL_INST;
+            }
+        }
+        return RISCV_EXCP_NONE;
+    }
+}
+
+static RISCVException fio(CPURISCVState *env, int csrno)
+{
+    AndesCsr *csr = &env->andes_csr;
+    if (get_field(csr->csrno[CSR_MMSC_CFG], MASK_MMSC_CFG_FIO) == 0) {
+        return RISCV_EXCP_ILLEGAL_INST;
+    }
     else {
         return RISCV_EXCP_NONE;
     }
 }
 
+static RISCVException veccfg(CPURISCVState *env, int csrno)
+{
+    AndesCsr *csr = &env->andes_csr;
+    if (riscv_has_ext(env, RVV)) {
+        if (get_field(csr->csrno[CSR_MMSC_CFG], MASK_MMSC_CFG_VECCFG) == 1) {
+            return RISCV_EXCP_NONE;
+        }
+    }
+    return RISCV_EXCP_ILLEGAL_INST;
+}
+
 static RISCVException pmnds(CPURISCVState *env, int csrno)
 {
     AndesCsr *csr = &env->andes_csr;
@@ -166,6 +204,13 @@ static RISCVException write_mecc_code(CPURISCVState *env, int csrno,
     return RISCV_EXCP_NONE;
 }
 
+static RISCVException write_ucode(CPURISCVState *env, int csrno,
+                                  target_ulong val)
+{
+    env->andes_csr.csrno[CSR_UCODE] = val & WRITE_MASK_CSR_UCODE;
+    return RISCV_EXCP_NONE;
+}
+
 static RISCVException write_uitb(CPURISCVState *env, int csrno,
                                  target_ulong val)
 {
@@ -202,6 +247,24 @@ static RISCVException write_mppib(CPURISCVState *env, int csrno,
     }
 }
 
+static RISCVException write_mfiob(CPURISCVState *env, int csrno,
+                                  target_ulong val)
+{
+    switch(riscv_cpu_mxl(env))
+    {
+        case MXL_RV32:
+            env->andes_csr.csrno[CSR_MFIOB] = val & WRITE_MASK_CSR_MFIOB_32;
+            return RISCV_EXCP_NONE;
+            break;
+        case MXL_RV64:
+            env->andes_csr.csrno[CSR_MFIOB] = val & WRITE_MASK_CSR_MFIOB_64;
+            return RISCV_EXCP_NONE;
+            break;
+        default:
+            return RISCV_EXCP_ILLEGAL_INST;
+    }
+}
+
 static RISCVException write_mpft_ctl(CPURISCVState *env, int csrno,
                                      target_ulong val)
 {
@@ -408,11 +471,11 @@ void andes_cpu_do_interrupt_post(CPUState *cs)
 riscv_csr_operations andes_csr_ops[CSR_TABLE_SIZE] = {
     /* ================== AndeStar V5 machine mode CSRs ================== */
     /* Configuration Registers */
-    [CSR_MICM_CFG]  = { "micm_cfg",          any, read_csr },
-    [CSR_MDCM_CFG]  = { "mdcm_cfg",          any, read_csr },
-    [CSR_MMSC_CFG]  = { "mmsc_cfg",          any, read_csr },
-    [CSR_MMSC_CFG2] = { "mmsc_cfg2",         any, read_csr },
-    [CSR_MVEC_CFG]  = { "mvec_cfg",          any, read_csr },
+    [CSR_MICM_CFG]  = { "micm_cfg",          any,    read_csr },
+    [CSR_MDCM_CFG]  = { "mdcm_cfg",          any,    read_csr },
+    [CSR_MMSC_CFG]  = { "mmsc_cfg",          any,    read_csr },
+    [CSR_MMSC_CFG2] = { "mmsc_cfg2",         any,    read_csr },
+    [CSR_MVEC_CFG]  = { "mvec_cfg",          veccfg, read_csr },
 
     /* Crash Debug CSRs */
     [CSR_MCRASH_STATESAVE]  = { "mcrash_statesave",  any, read_csr },
@@ -431,7 +494,7 @@ riscv_csr_operations andes_csr_ops[CSR_TABLE_SIZE] = {
     [CSR_MCCTLCOMMAND]   = { "mcctlcommand",      any, read_csr, write_csr  },
     [CSR_MCCTLDATA]      = { "mcctldata",         any, read_csr, write_csr  },
     [CSR_MPPIB]          = { "mppib",             ppi, read_csr, write_mppib},
-    [CSR_MFIOB]          = { "mfiob",             any, read_csr, write_csr  },
+    [CSR_MFIOB]          = { "mfiob",             fio, read_csr, write_mfiob},
 
     /* Hardware Stack Protection & Recording */
     [CSR_MHSP_CTL]     = { "mhsp_ctl",            hsp, read_csr,
@@ -539,8 +602,8 @@ riscv_csr_operations andes_csr_ops[CSR_TABLE_SIZE] = {
 
     /* =================== AndeStar V5 user mode CSRs =================== */
     /* User mode control registers */
-    [CSR_UITB]           = { "uitb",              ecd, read_csr, write_uitb},
-    [CSR_UCODE]          = { "ucode",             any, read_csr, write_csr},
+    [CSR_UITB]           = { "uitb",              ecd,  read_csr, write_uitb },
+    [CSR_UCODE]          = { "ucode",             edsp, read_csr, write_ucode},
     [CSR_UDCAUSE]        = { "udcause",           any, read_csr, write_csr},
     [CSR_UCCTLBEGINADDR] = { "ucctlbeginaddr",    any, read_csr, write_csr},
     [CSR_UCCTLCOMMAND]   = { "ucctlcommand",      any, read_csr, write_csr},
diff --git a/target/riscv/csr_andes.h b/target/riscv/csr_andes.h
index 9f17d56dcf..a307588b65 100644
--- a/target/riscv/csr_andes.h
+++ b/target/riscv/csr_andes.h
@@ -105,6 +105,10 @@ typedef struct AndesVec {
 #define WRITE_MASK_CSR_SLIEP                0x10F0000
 #define WRITE_MASK_CSR_MPPIB_32             0xFFFFFC3F
 #define WRITE_MASK_CSR_MPPIB_64             0xFFFFFFFFFFFFFC3F
+#define WRITE_MASK_CSR_MFIOB_32             0xFFFFFC3F
+#define WRITE_MASK_CSR_MFIOB_64             0xFFFFFFFFFFFFFC3F
+#define WRITE_MASK_CSR_UCODE                0x1
+#define WRITE_MASK_CSR_MVEC_CFG             0x3FFFF
 
 void andes_csr_init(AndesCsr *);
 void andes_vec_init(AndesVec *);
-- 
2.34.1

